src/neo/SmartContract/Native/Designate/DesignateContract.cs

```diff
@@ -37,34 +37,72 @@ public ECPoint[] GetDesignatedByRole(StoreView snapshot, Role role)
         {
             if (!Enum.IsDefined(typeof(Role), role))
                 throw new ArgumentOutOfRangeException(nameof(role));
-            return snapshot.Storages[CreateStorageKey((byte)role)].GetInteroperable<NodeList>().ToArray();
+            NodeMap map = snapshot.Storages[CreateStorageKey((byte)role)].GetInteroperable<NodeMap>();
+            List<uint> keys = map.Keys.ToList();
+            if (keys.Count == 0) return System.Array.Empty<ECPoint>();
+            return map[keys.Last()].ToArray();
+        }
+
+        [ContractMethod(0_01000000, CallFlags.AllowStates)]
+        public ECPoint[] GetDesignatedByRoleAndIndex(StoreView snapshot, Role role, uint index)
+        {
+            if (!Enum.IsDefined(typeof(Role), role))
+                throw new ArgumentOutOfRangeException(nameof(role));
+            if (snapshot.Height + 2 < index)
+                throw new ArgumentOutOfRangeException(nameof(index));
+            NodeMap map = snapshot.Storages[CreateStorageKey((byte)role)].GetInteroperable<NodeMap>();
+            List<uint> keys = map.Keys.ToList();
+            if (keys.Count == 0) return System.Array.Empty<ECPoint>();
+
+            uint height = 0;
+            foreach (uint h in keys)
+            {
+                if (index <= h) break;
+                height = h;
+            }
+
+            if (0 < height)
+                return map[height].ToArray();
+
+            return System.Array.Empty<ECPoint>();
         }
 
         [ContractMethod(0, CallFlags.AllowModifyStates)]
-        private void DesignateAsRole(ApplicationEngine engine, Role role, ECPoint[] nodes)
+        private bool DesignateAsRole(ApplicationEngine engine, Role role, ECPoint[] nodes)
         {
             if (nodes.Length == 0 || nodes.Length > 32)
                 throw new ArgumentException();
             if (!Enum.IsDefined(typeof(Role), role))
                 throw new ArgumentOutOfRangeException(nameof(role));
             if (!CheckCommittee(engine)) throw new InvalidOperationException();
-            NodeList list = engine.Snapshot.Storages.GetAndChange(CreateStorageKey((byte)role)).GetInteroperable<NodeList>();
-            list.Clear();
-            list.AddRange(nodes);
+            NodeMap map = engine.Snapshot.Storages.GetAndChange(CreateStorageKey((byte)role)).GetInteroperable<NodeMap>();
+            List<ECPoint> list = nodes.ToList();
             list.Sort();
+            map[engine.Snapshot.Height + 1] = list;
+            return true;
         }
 
-        private class NodeList : List<ECPoint>, IInteroperable
+        private class NodeMap : Dictionary<uint, List<ECPoint>>, IInteroperable
```

Comment:
If we store without limits, we can produce a DoS in the future.

Reply:
We can store in different key.

Reply:
Done.